src/pharmpy/tools/allometry/tool.py

```diff
@@ -113,6 +117,75 @@ def _add_allometry_on_model(
     return model
 
 
+@same_signature_as(create_workflow)
+def validate_input(
+    model,
+    allometric_variable,
+    reference_value,
+    parameters,
+    initials,
+    lower_bounds,
+    upper_bounds,
+    fixed,
+):
+    if not isinstance(allometric_variable, (str, sympy.Expr)):
+        raise TypeError(
+            f'Invalid allometric_variable: got "{allometric_variable}"'
+            f' of type {type(allometric_variable)}, must be a str/sympy.Expr.'
+        )
+
+    if not isinstance(reference_value, (str, int, float, sympy.Expr)):
+        raise TypeError(
+            f'Invalid reference_value: got "{reference_value}"'
+            f' of type {type(reference_value)}, must be a str/int/float/sympy.Expr.'
+        )
+
+    if not isinstance(parameters, (type(None), list)):
+        raise TypeError(
+            f'Invalid parameters: got "{parameters}" of type {type(parameters)}, must be None/NULL or a list.'
+        )
+
+    if not isinstance(initials, (type(None), list)):
+        raise TypeError(
+            f'Invalid initials: got "{initials}" of type {type(initials)}, must be None/NULL or a list.'
+        )
+    if not isinstance(lower_bounds, (type(None), list)):
+        raise TypeError(
+            f'Invalid lower_bounds: got "{lower_bounds}" of type {type(lower_bounds)}, must be None/NULL or a list.'
+        )
+    if not isinstance(upper_bounds, (type(None), list)):
+        raise TypeError(
+            f'Invalid upper_bounds: got "{upper_bounds}" of type {type(upper_bounds)}, must be None/NULL or a list.'
+        )
+
+    if not isinstance(fixed, bool):
+        raise TypeError(f'Invalid fixed: got "{fixed}" of type {type(fixed)}, must be a bool.')
+
```

Comment:
Yep, that's a good idea. Maybe with another decorator. I'll see what can be done.

Reply:
There is [`enforce`](https://github.com/RussBaz/enforce) but looks very much unmaintained.

Reply:
There is also [`type_enforced`](https://github.com/connor-makowski/type_enforced) but looks young.

Reply:
The fact that https://github.com/python/typing/issues/769 is still open makes avoiding duplication more difficult.

Reply:
Yeah, but is an external dependency needed? All that functionality is in `inspect` or?

Reply:
`inspect` allows you to read the annotations, but not check that this are respected at runtime. I am working on an adhoc solution since I could not find a satisfactory candidate for dependency.

Reply:
On the idea of relying on `isinstance` only:
  1. It does not work reliably: https://docs.python.org/3/library/collections.abc.html#id18
  2. It only checks the root of the type tree. So it will not check that a `Collection[int]` contains instances of `int`.

Reply:
`parameter.annotation` (or `fn.__annotations__[param.name]`) will not work in the future because of [PEP 563](https://peps.python.org/pep-0563/). We need to use `get_type_hints(fn)`.